RecoLocalMuon/GEMSegment/plugins/GEMSegmentAlgorithm.cc

```diff
@@ -0,0 +1,474 @@
+   	                 	     << " -- "<<rhID.rawId()<<" = "<<rhID<<" ]";
+    #endif
+  }  
+
+  for(size_t iRH_old = 0;iRH_old<oldChain.size();++iRH_old){
+    GEMDetId rhID                  = oldChain[iRH_old]->gemId();
+    const GEMEtaPartition * rhEP   = (theEnsemble.second.find(rhID.rawId()))->second;
+    if(!rhEP) throw cms::Exception("GEMEtaPartition not found") << "Corresponding GEMEtaPartition to GEMDetId: "<<rhID<<" not found in the GEMEnsemble";
+    LocalPoint rhLP_inEtaPartFrame = oldChain[iRH_old]->localPosition();
+    GlobalPoint rhGP_inCMSFrame    = rhEP->toGlobal(rhLP_inEtaPartFrame);
+    layer_old.push_back((rhID.station()-1)*2+rhID.layer());
+    // std::vector<int> layer_tmp; layer_tmp.push_back((rhID.station()-1)*2+rhID.layer()); layer_old.push_back(layer_tmp);
+    phi_old.push_back(rhGP_inCMSFrame.phi());
+    eta_old.push_back(rhGP_inCMSFrame.eta());
+    bx_old.push_back(oldChain[iRH_old]->BunchX());
+    // Log Message
+    #ifdef EDM_ML_DEBUG // have lines below only compiled when in debug mode
+    const GEMSuperChamber * rhCH   = theEnsemble.first;
+    LocalPoint rhLP_inChamberFrame = rhCH->toLocal(rhGP_inCMSFrame);
+    edm::LogVerbatim("GEMSegmentAlgorithm") << "[GoodToMerge::Old Chain][RecHit :: Loc x = "<<std::showpos<<std::setw(9)<<rhLP_inChamberFrame.x()<<" Loc y = "<<std::showpos<<std::setw(9)<<rhLP_inChamberFrame.y()
+                                     << " -- Glob eta = "<<std::showpos<<std::setw(9)<<rhGP_inCMSFrame.eta()<<" Glob phi = "<<std::showpos<<std::setw(9)<<rhGP_inCMSFrame.phi()
+                                     << " -- BX = "<<oldChain[iRH_old]->BunchX()
+                                     << " -- "<<rhID.rawId()<<" = "<<rhID<<" ]";
+    #endif
+  }
+
+  // Different Requirements: 
+  bool phiRequirementOK   = false; // once it is true in the loop, it is ok to merge
+  bool etaRequirementOK   = false; // once it is true in the loop, it is ok to merge
+  bool bxRequirementOK    = false; // once it is true in the loop, it is ok to merge
+  bool layerRequirementOK = true;  // once it is false in the loop, it is not ok to merge
+
+  for(size_t jRH_new = 0; jRH_new<phi_new.size(); ++jRH_new){
+    for(size_t jRH_old = 0; jRH_old<phi_old.size(); ++jRH_old){
+
+      // Eta & Phi
+      // to be chained, two hits need also to be "close" in phi and eta
+      // allow only change false --> true. once true, leave it true
+      if(phiRequirementOK==false) phiRequirementOK = std::abs(reco::deltaPhi(phi_new[jRH_new],phi_old[jRH_old])) < dPhiChainBoxMax;
+      if(etaRequirementOK==false) etaRequirementOK = fabs(eta_new[jRH_new]-eta_old[jRH_old]) < dEtaChainBoxMax;
+
+      // Layers
+      // hits on the same layer should not be allowed ==> if abs(layer_new - layer_old) > 0 is ok. if = 0 is false
+      // allow only change true --> false. once false, leave it false
+      if(layerRequirementOK==true) { 
+	layerRequirementOK = layer_new[jRH_new] != layer_old[jRH_old];
+      }
+
```

Comment:
or better just remove layerRequirementOK completely and use

if ( layer_new[jRH_new] == layer_old[jRH_old] ) return false;